k2/csrc/fsa_equivalent.cc

```diff
@@ -62,8 +90,63 @@ bool IsRandEquivalent(const Fsa &a, const Fsa &b, std::size_t npath /*=100*/) {
   return true;
 }
 
-bool RandomPath(const Fsa &a, Fsa *b,
-                std::vector<int32_t> *state_map /*=nullptr*/) {
+template <class FbWeightType>
+bool IsRandEquivalent(const Fsa &a, const float *a_weights, const Fsa &b,
+                      const float *b_weights, std::size_t npath /*= 100*/) {
+  Fsa connected_a, connected_b, valid_a, valid_b;
+  std::vector<int32_t> connected_a_arc_map, connected_b_arc_map,
+      valid_a_arc_map, valid_b_arc_map;
+  Connect(a, &connected_a, &connected_a_arc_map);
+  Connect(b, &connected_b, &connected_b_arc_map);
+  ArcSort(connected_a, &valid_a, &valid_a_arc_map);  // required by `intersect`
+  ArcSort(connected_b, &valid_b, &valid_b_arc_map);
+  if (IsEmpty(valid_a) && IsEmpty(valid_b)) return true;
+  if (IsEmpty(valid_a) || IsEmpty(valid_b)) return false;
+
+  // Get arc weights
+  std::vector<float> valid_a_weights(valid_a.arcs.size());
+  std::vector<float> valid_b_weights(valid_b.arcs.size());
+  GetArcWeights(a_weights, connected_a_arc_map, valid_a_arc_map,
+                &valid_a_weights);
+  GetArcWeights(b_weights, connected_b_arc_map, valid_b_arc_map,
+                &valid_b_weights);
+
+  // Check that arc labels are compatible.
+  std::unordered_set<int32_t> labels_a, labels_b, labels_difference;
+  for (const auto &arc : valid_a.arcs) labels_a.insert(arc.label);
+  for (const auto &arc : valid_b.arcs) labels_b.insert(arc.label);
+  SetDifference(labels_a, labels_b, &labels_difference);
+  if (labels_difference.size() >= 2 ||
+      (labels_difference.size() == 1 &&
+       (*(labels_difference.begin()) != kEpsilon)))
+    return false;
+
+  std::random_device rd;
+  std::mt19937 gen(rd());
+  std::bernoulli_distribution coin(0.5);
+  for (auto i = 0; i != npath; ++i) {
+    const auto &fsa = coin(gen) ? valid_a : valid_b;
+    Fsa path, valid_path;
+    RandomPathWithoutEpsilonArc(fsa, &path);  // path is already connected
+    ArcSort(path, &valid_path);
+
+    Fsa a_compose_path, b_compose_path;
+    std::vector<float> a_compose_weights, b_compose_weights;
+    Intersect(valid_a, valid_a_weights.data(), path, &a_compose_path,
+              &a_compose_weights);
+    Intersect(valid_b, valid_b_weights.data(), path, &b_compose_path,
+              &b_compose_weights);
+    double sum_a = ShortestDistance<FbWeightType>(a_compose_path,
+                                                  a_compose_weights.data());
+    double sum_b = ShortestDistance<FbWeightType>(b_compose_path,
+                                                  b_compose_weights.data());
+    if (!DoubleApproxEqual(sum_a, sum_b)) return false;
+  }
+  return true;
+}
+
+bool RandomPath(const Fsa &a, Fsa *b, bool no_epsilon_arc,
```

Comment:
This overload is not declared in the header file `fsa_equivalent.h`.

Reply:
It would be great if this function is placed in the above unnamed namespace
if it is used only inside this `.cc` file.

Reply:
BTW, if something is used only in a .cc file I'd rather it be `static` to clarify that.

Reply:
Sure, will move to unnamed namespace and add `static`